nav2_toolkit/include/nav2_toolkit/pose_saver_node.hpp

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2025 Pranav Kolekar
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_TOOLKIT__POSE_SAVER_NODE_HPP_
+#define NAV2_TOOLKIT__POSE_SAVER_NODE_HPP_
+
+#include <yaml-cpp/yaml.h>
+#include <memory>
+#include <string>
+#include <rclcpp/rclcpp.hpp>
+#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
+#include <nav2_msgs/srv/set_initial_pose.hpp>
+#include <std_srvs/srv/trigger.hpp>
+#include "nav2_ros_common/service_client.hpp"
+
+namespace nav2_toolkit
+{
+
+/**
+ * @class PoseSaverNode
+ * @brief A utility node to periodically save and restore the robot's pose from a YAML file
+ *
+ * This node subscribes to the robot's pose topic (typically `amcl_pose`), stores the latest
+ * pose to a YAML file periodically, and provides services to control pose saving/restoration.
+ * The node can automatically restore the last known pose on startup and provides manual
+ * restoration through service calls.
+ *
+ * Key Features:
+ * - Periodic pose saving to YAML file with atomic write operations
+ * - Automatic pose restoration on node startup (configurable)
+ * - Service-based control for starting/stopping pose saving
+ * - Manual pose restoration via service call
+ * - Crash-safe file operations to prevent data corruption
+ *
+ * Services Provided:
+ * - `start_pose_saver`: Start periodic pose saving
+ * - `stop_pose_saver`: Stop periodic pose saving
+ * - `localise_at_last_known_position`: Restore the saved pose via service call
+ *
+ * Topics Subscribed:
+ * - `amcl_pose` (geometry_msgs::msg::PoseWithCovarianceStamped): Robot pose updates
+ *
+ * Service Clients:
+ * - `/initialpose` (nav2_msgs::srv::SetInitialPose): To restore saved poses
+ */
+class PoseSaverNode : public rclcpp::Node
+{
+public:
+  /**
+   * @brief Constructor for PoseSaverNode
+   * @param options Node options used for configuring the ROS node
+   */
+  explicit PoseSaverNode(const rclcpp::NodeOptions & options);
+
+protected:
+  // === Callbacks ===
+
+  /**
+   * @brief Callback to store incoming robot pose messages
+   * @param msg Shared pointer to the received pose message with covariance
+   */
+  void pose_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg);
```

Comment:
The methods in this class do not match the styling of the repository. Methods are lower case first letter and otherwise camelCase.